osgar/ro2018.py

```diff
@@ -134,6 +164,10 @@ def listen(self):
     def publish(self, channel, data):
```

Comment:
This function has several issues. First, it ignores its `channel` argument (does not assert it is the same channel as during record, it accepts any output). Second, when `self.outputs` are empty (so the code should never call `publish` during replay) and the code calls `publish`, this call is silently ignored. Also I somehow expected that `self.outputs` would have the same meaning as during recording but the map is inverted - while that is ok, it surprised me.

A possible improvement:

```python
def publish(self, channel, data):
    assert channel in self.outputs.values()
    dt, stream_id, raw_data = next(self.reader)
    while stream_id not in self.outputs:
        assert stream_id in self.inputs, stream_id
        self.buffer_queue.put((dt, stream_id, raw_data))
        dt, stream_id, raw_data = next(self.reader)
    assert channel == self.outputs[stream_id]
    ref_data = literal_eval(raw_data.decode('ascii'))
    assert data == ref_data, (data, ref_data)
```

Known limitation (modulo typos) - it will correctly replay only single threaded modules where the order of `publish` calls is deterministic.